### PR TITLE
add option to edit intros/credits

### PR DESCRIPTION
**Pull request by jumoog**

### CODE REVIEW

File: ConfusedPolarBear.Plugin.IntroSkipper/Controllers/VisualizationController.cs [hunk 1]
```diff
@@ -157,18 +158,48 @@ public ActionResult EraseSeason([FromRoute] string series, [FromRoute] string se
...
+    /// <param name="id">Episode ID to update timestamps for.</param>
+    /// <param name="timestamps">New timestamps Introduction/Credits start and end times.</param>
+    /// <response code="204">New timestamps saved.</response>
+    /// <returns>No content.</returns>
+    [HttpPost("Episode/{Id}/UpdateTimestamps")]
+    public ActionResult UpdateTimestamps([FromRoute] Guid id, [FromBody] TimeStamps timestamps)
+    {
+        if (timestamps?.Introduction.IntroStart > 0.0 && timestamps?.Introduction.IntroEnd > 0.0)
+        {
+            var tr = new TimeRange(timestamps.Credits.IntroStart, timestamps.Credits.IntroEnd);
```
Comment (rlauuzo) [inline ConfusedPolarBear.Plugin.IntroSkipper/Controllers/VisualizationController.cs:192]: timestamps.Introduction

**Comment by jumoog** [comment]
after updating the credits end value is -2

**Comment by jumoog** [comment]
the value is SecondsOfIntroToPlay

**Comment by AbandonedCart** [comment]
Isn't skipping the credits now done by going 10 seconds over the end of the video?

**Comment by jumoog** [comment]
before saving
```json
{
    "Introduction": {
        "EpisodeId": "c3a3cb575744d18b362c91cb73325a4c",
        "Valid": true,
        "IntroStart": 335,
        "IntroEnd": 364,
        "ShowSkipPromptAt": 335,
        "HideSkipPromptAt": 364
    },
    "Credits": {
        "EpisodeId": "c3a3cb575744d18b362c91cb73325a4c",
        "Valid": true,
        "IntroStart": 2875,
        "IntroEnd": 2998,
        "ShowSkipPromptAt": 2875,
        "HideSkipPromptAt": 2998
    }
}
```

```xml
<Intro>
  <EpisodeId>c3a3cb57-5744-d18b-362c-91cb73325a4c</EpisodeId>
  <HideSkipPromptAt>0</HideSkipPromptAt>
  <IntroEnd>366</IntroEnd>
  <IntroStart>335</IntroStart>
  <ShowSkipPromptAt>0</ShowSkipPromptAt>
</Intro>
```

**Comment by jumoog** [comment]
after saving:
```xml
<Intro>
  <EpisodeId>c3a3cb57-5744-d18b-362c-91cb73325a4c</EpisodeId>
  <HideSkipPromptAt>0</HideSkipPromptAt>
  <IntroEnd>364</IntroEnd>
  <IntroStart>335</IntroStart>
  <ShowSkipPromptAt>0</ShowSkipPromptAt>
</Intro>
```

**Comment by jumoog** [comment]
new API Endpoint: `Episode/{Id}/Timestamps`

HTTP POST to update Timestamps.
HTTP GET to receive the unmodified Timestamps. If Intro/Outro not exists the API returns 0

**Comment by jumoog** [comment]
my frontend skills sucks so it looks kinda ugly now:

grafik

**Comment by jumoog** [comment]
6qlw03

**Comment by AbandonedCart** [comment]
The reason I asked is if the code is set to throw 10 seconds onto the end and let’s say I back the end up 20 seconds, wouldn’t it be 10 seconds later than I expected?

**Comment by jumoog** [comment]
> The reason I asked is if the code is set to throw 10 seconds onto the end and let’s say I back the end up 20 seconds, wouldn’t it be 10 seconds later than I expected?

Yes. That's why I now use the unmodified times and let the plugin do its magic in the background during playback.

**Comment by rlauuzo** [comment]
The 10 seconds are only added if the credits end within the last 3 seconds. Because 2 seconds are always subtracted from the end, and there is some rounding down happening for the duration.

**Comment by jumoog** [comment]
chrome has time picker
grafik

**Comment by AbandonedCart** [comment]
> The 10 seconds are only added if the credits end within the last 3 seconds. Because 2 seconds are always subtracted from the end, and there is some rounding down happening for the duration.

If one is static and the other is dynamic, they shouldn’t collide, but that one rare edge case seems to be the theme of the issues page.

**Comment by AbandonedCart** [comment]
> chrome has time picker

If you’re using input type = time, then Firefox and I believe Safari also work.

**Comment by jumoog** [comment]
> > chrome has time picker
> 
> If you’re using input type = time, then [Firefox](https://blog.nightly.mozilla.org/2017/06/12/datetime-inputs-enabled-on-nightly/) and I believe Safari also work.
> 
> Supported after iOS 5.0 for Safari mobile and IE after version 12.

firefox has no direct picker you can edit the hours/minutes/seconds

**Comment by AbandonedCart** [comment]
> firefox has no direct picker you can edit the hours/minutes/seconds

I'm on firefox. It support the time input type.

**Comment by jumoog** [comment]
I think i'll keep it that way and do a config cleanup in a seperate PR
grafik

**Comment by jumoog** [comment]
i hate frontend 😂

**Comment by jumoog** [comment]
ok final version
grafik

**Comment by jumoog** [comment]
ready for review

File: ConfusedPolarBear.Plugin.IntroSkipper/Controllers/VisualizationController.cs [hunk 2]
```diff
@@ -157,18 +158,21 @@ public ActionResult EraseSeason([FromRoute] string series, [FromRoute] string se
...
     /// <response code="204">New introduction timestamps saved.</response>
     /// <returns>No content.</returns>
     [HttpPost("Episode/{Id}/UpdateIntroTimestamps")]
-    public ActionResult UpdateTimestamps([FromRoute] Guid id, [FromBody] Intro timestamps)
+    public ActionResult UpdateIntroTimestamps([FromRoute] Guid id, [FromBody] Intro timestamps)
     {
-        var tr = new TimeRange(timestamps.IntroStart, timestamps.IntroEnd);
-        Plugin.Instance!.Intros[id] = new Intro(id, tr);
-        Plugin.Instance.SaveTimestamps(AnalysisMode.Introduction);
+        if (timestamps.IntroStart > 0.0 && timestamps.IntroEnd > 0.0)
```
Comment (rlauuzo) [inline ConfusedPolarBear.Plugin.IntroSkipper/Controllers/VisualizationController.cs:170]:
```suggestion
        if (timestamps.IntroEnd > 0.0)
```

File: ConfusedPolarBear.Plugin.IntroSkipper/Controllers/SkipIntroController.cs [hunk 3]
```diff
@@ -48,6 +49,58 @@ public SkipIntroController()
...
+            Plugin.Instance!.Intros[id] = new Intro(id, tr);
+        }
+
+        if (timestamps?.Credits.IntroEnd > 0.0)
+        {
+            var cr = new TimeRange(timestamps.Credits.IntroStart, timestamps.Credits.IntroEnd);
+            Plugin.Instance!.Credits[id] = new Intro(id, cr);
+        }
+
+        Plugin.Instance?.SaveTimestamps(AnalysisMode.Introduction);
```
Comment (rlauuzo) [inline ConfusedPolarBear.Plugin.IntroSkipper/Controllers/SkipIntroController.cs:75]:
```suggestion
        Plugin.Instance?.SaveTimestamps(AnalysisMode.Introduction);
        Plugin.Instance!.SaveTimestamps(AnalysisMode.Credits);
```

**Comment by jumoog** [comment]
should we mark UpdateIntroTimestamps / Episode/{Id}/UpdateIntroTimestamps as deprecated?

**Comment by jumoog** [comment]
lgtm

**Comment by AbandonedCart** [comment]
Did you try it out on mobile? Asking because each line looks like a decent amount of content

**Comment by jumoog** [comment]
yes works but could be better

**Pull request merged**